types.go

```diff
@@ -168,25 +172,46 @@ func (fs *fields) OmitEmpty(strct reflect.Value, forced bool) []*field {
 	return fields
 }
 
+func (fs *fields) OrderArray(forced bool) ([]*field, error) {
+	if !fs.AsOrderedArray && !forced {
+		return fs.List, nil
+	}
+
+	maxOrder := 0
+	fsMap := make(map[int]*field)
+	for _, f := range fs.List {
+		order := getFieldOrder(f.name)
+		if _, found := fsMap[order]; found {
+			return nil, fmt.Errorf("msgpack: %s has repeated order for field %s", fs.Type, f.name)
+		}
+		fsMap[order] = f
+		if order > maxOrder {
+			maxOrder = order
+		}
+	}
+
+	fsOrdered := make([]*field, maxOrder+1)
+	for order, f := range fsMap {
+		fsOrdered[order] = f
+	}
+	return fsOrdered, nil
```

Comment:
This should be done once - not each time the struct is decoded. It is rather expensive to allocate a slice + a map + parse field names.